src/main/webapp/IFrame.js

```diff
@@ -0,0 +1,151 @@
+  player = new YT.Player('ytplayer', { // eslint-disable-line no-undef
+    events: {
+      'onStateChange': onPlayerStateChange,
+      'onPlaybackRateChange': onPlayerPlaybackRateChange,
+    },
+  });
+}
+
+function sendInfo(playing) {
+  const xhttp = new XMLHttpRequest();
+  xhttp.onreadystatechange = function() {
+    if (this.readyState == 4 && this.status == 200) {
+      console.log('Request sent');
+    }
+  };
+  xhttp.open('POST', '/Playback-Test', true);
+  xhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
+  const time = player.getCurrentTime();
+  const speed = player.getPlaybackRate();
+  xhttp.send('timestamp=' + time + '&videoSpeed=' + speed + '&isPlaying='+
+    playing);
+}
+
+let timeout;
+function onPlayerStateChange() {
+  console.log(player.getPlayerState());
+  if (!videoUpdating) {
+    switch (player.getPlayerState()) {
+      case 1: // Playing
+        clearTimeout(timeout);
+        sendInfo(true);
+        break;
+      case 2: // paused
+        timeout = setTimeout(sendInfo, 100, false);
+        break;
+      case 3: // Buffering
+        clearTimeout(timeout);
+        break;
+      case -1: // Just before video starts
+        beginFetchingLoop();
+        break;
+      case 0: // Ended
+        endFetchingLoop();
+    }
+  }
+}
+
+// Get the time of the current video
+function getTime() { // eslint-disable-line no-unused-vars
+  console.log(player.getCurrentTime());
+}
+
+function setTime() { // eslint-disable-line no-unused-vars
+  player.seekTo(60, true);
+}
+
+function onPlayerPlaybackRateChange() {
+  if (!videoUpdating) sendInfo(player.getPlayerState() === 1);
+}
+
+function halfSpeed() { // eslint-disable-line no-unused-vars
+  player.setPlaybackRate(0.5);
+}
+
+function doubleSpeed() { // eslint-disable-line no-unused-vars
+  player.setPlaybackRate(2);
+}
+
+function normalSpeed() { // eslint-disable-line no-unused-vars
+  player.setPlaybackRate(1);
+}
+
+function fetchData() {
+  const request = new XMLHttpRequest();
+  request.onreadystatechange = function() {
+    if (this.readyState == 4 && this.status == 200) {
+      updateVideo(this.responseText);
+    }
+  };
+  request.open('GET', '/Playback-Test', true);
+  request.send();
+}
+
+// return true if player time is within 5 seconds of master time
+function timesInRange(serverVidTime) {
+  return Math.abs(player.getCurrentTime() - serverVidTime) < 3;
+}
+
+// return true if player state is different than master state
+function differentStates(serverVidIsPlaying) {
+  if (player.getPlayerState() != 1) { // player paused
+    if (serverVidIsPlaying) { // master playing
+      return true;
+    }
+    return false;
+  }
+  if (!serverVidIsPlaying) {
+    return true;
+  }
+  return false;
+}
+
+function updateVideo(text) {
+  console.log(text);
+  const videoInfo = JSON.parse(text);
+  videoUpdating = true;
+  if (!timesInRange(videoInfo.timestamp)) {
+    player.seekTo(videoInfo.timestamp - 0.75, true);
```

Comment:
curious, why the 0.75? Can we move it to a constant to understand what it means.

Reply:
Sure, the way I implemented updating the video timestamp on the master version stored in the datastore was by increasing the video time by 0.75 seconds on every fetch request (I've been testing it on two screens, which each make fetch requests once per 1.5 seconds, so after they've each made fetch requests (2 * 0.75), the video has added the appropriate 1.5 seconds). However, when one of the videos needs to be updated (i.e. moved to a new time) this causes a problem because I've added time to the timestamp I was supposed to be updating to. For example, if the other person sent "go to 30 seconds", the updating client receives "go to 30.75 seconds" (or even 31.5 seconds in the bad case where the other person's client fetched sooner than this one). By subtracting the 0.75, I at least keep this offset within one second. I'll give it a constant name to be clearer.